engine/server/session.go

```diff
@@ -0,0 +1,953 @@
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"runtime"
+	"runtime/debug"
+	"sync"
+	"time"
+
+	"dagger.io/dagger/telemetry"
+	"github.com/99designs/gqlgen/graphql/handler"
+	"github.com/Khan/genqlient/graphql"
+	"github.com/koron-go/prefixw"
+	"github.com/moby/buildkit/cache/remotecache"
+	bkclient "github.com/moby/buildkit/client"
+	bkfrontend "github.com/moby/buildkit/frontend"
+	bkgw "github.com/moby/buildkit/frontend/gateway/client"
+	bksession "github.com/moby/buildkit/session"
+	bksolver "github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/moby/buildkit/util/progress/progressui"
+	"github.com/sirupsen/logrus"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/dagger/dagger/analytics"
+	"github.com/dagger/dagger/auth"
+	"github.com/dagger/dagger/core"
+	"github.com/dagger/dagger/core/schema"
+	"github.com/dagger/dagger/dagql"
+	"github.com/dagger/dagger/dagql/call"
+	"github.com/dagger/dagger/engine"
+	"github.com/dagger/dagger/engine/buildkit"
+	"github.com/dagger/dagger/engine/cache"
+	"github.com/dagger/dagger/engine/slog"
+	enginetel "github.com/dagger/dagger/engine/telemetry"
+)
+
+type daggerSession struct {
+	sessionID          string
+	mainClientCallerID string
+
+	state   daggerSessionState
+	stateMu sync.RWMutex
+
+	clients  map[string]*daggerClient // clientID -> client
+	clientMu sync.RWMutex
+
+	// the http endpoints being served (as a map since APIs like shellEndpoint can add more)
+	endpoints  map[string]http.Handler
+	endpointMu sync.RWMutex
+
+	services *core.Services
+
+	analytics analytics.Tracker
+
+	secretStore  *core.SecretStore
+	authProvider *auth.RegistryAuthProvider
+
+	cacheExporterCfgs []bkgw.CacheOptionsEntry
+	cacheImporterCfgs []bkgw.CacheOptionsEntry
+
+	refs   map[buildkit.Reference]struct{}
+	refsMu sync.Mutex
+
+	containers   map[bkgw.Container]struct{}
+	containersMu sync.Mutex
+
+	dagqlCache dagql.Cache
+}
+
+type daggerSessionState string
+
+const (
+	sessionStateUninitialized daggerSessionState = ""
```

Comment:
Any reason not to make this (and `clientStateUninitialized`) more explicitly `"uninitialized"` or something? Might be nice to have clarity on it vs. a value being dropped somewhere (assuming this ends up in logs)

Reply:
Good point, the fact that the zero value was uninitialized was mildly convenient but having it be more explicit is better I agree. Fortunately there is exactly one place session+client get initialized, so easy to just explicitly set there.